src/crimson/os/seastore/journal.h

```diff
@@ -245,19 +219,168 @@ class Journal {
+    std::optional<seastar::shared_promise<
+        std::optional<journal_seq_t> > > io_promise;
+  };
+
+  class RecordSubmitter {
+    enum class state_t {
+      IDLE = 0, // outstanding_io == 0
+      PENDING,  // outstanding_io <  io_depth_limit
+      FULL      // outstanding_io == io_depth_limit
+      // OVERFLOW: outstanding_io >  io_depth_limit is impossible
+    };
+
+  public:
+    RecordSubmitter(std::size_t io_depth, JournalSegmentManager&);
+
+    void set_write_pipeline(WritePipeline *_write_pipeline) {
+      write_pipeline = _write_pipeline;
+    }
+
+    using submit_ret = Journal::submit_record_ret;
+    submit_ret submit(record_t&&, OrderingHandle&);
+
+  private:
+    void update_state();
+
+    void increment_io() {
+      ++num_outstanding_io;
+      update_state();
+    }
+
+    void decrement_io_with_flush() {
+      assert(num_outstanding_io > 0);
+      --num_outstanding_io;
+#ifndef NDEBUG
+      auto prv_state = state;
+#endif
+      update_state();
+
+      if (wait_submit_promise.has_value()) {
+        assert(prv_state == state_t::FULL);
+        wait_submit_promise->set_value();
+        wait_submit_promise.reset();
+      }
+
+      if (!p_current_batch->is_empty()) {
+        flush_current_batch();
+      }
+    }
+
+    void pop_free_batch() {
+      assert(p_current_batch == nullptr);
+      assert(!free_batch_ptrs.empty());
+      p_current_batch = free_batch_ptrs.front();
+      assert(p_current_batch->is_empty());
+      assert(p_current_batch == &batches[p_current_batch->get_index()]);
+      free_batch_ptrs.pop_front();
+    }
+
+    void finish_submit_batch(RecordBatch*, std::optional<journal_seq_t>);
+
+    void flush_current_batch();
+
+    seastar::future<std::pair<paddr_t, journal_seq_t>>
+    mark_record_committed_in_order(
+        OrderingHandle&, const journal_seq_t&, const record_size_t&);
+
+    using submit_record_ertr = JournalSegmentManager::write_ertr;
+    using submit_record_ret = submit_record_ertr::future<
+      std::pair<paddr_t, journal_seq_t> >;
+    submit_record_ret submit_record(
+        record_t&&, const record_size_t&, OrderingHandle& handle);
+
+    using submit_pending_ret = submit_record_ret;
+    submit_pending_ret submit_pending(
+        record_t&&, const record_size_t&, OrderingHandle &handle, bool flush);
+
+    using do_submit_ret = submit_record_ret;
+    do_submit_ret do_submit(
+        record_t&&, const record_size_t&, OrderingHandle&);
+
+    state_t state = state_t::IDLE;
+    std::size_t num_outstanding_io = 0;
+    std::size_t io_depth_limit;
+
+    WritePipeline* write_pipeline = nullptr;
+    JournalSegmentManager& journal_segment_manager;
+    std::unique_ptr<RecordBatch[]> batches;
+    std::size_t current_batch_index;
+    // should not be nullptr after constructed
+    RecordBatch* p_current_batch = nullptr;
+    seastar::circular_buffer<RecordBatch*> free_batch_ptrs;
+    std::optional<seastar::shared_promise<> > wait_submit_promise;
```

Comment:
Looks to me a plain promise here should be suffice.

Reply:
Ah, right since we don't need more than once reference holder for the promise itself.

Reply:
fixed.